src/BSON/Binary.c

```diff
@@ -297,14 +328,339 @@ void php_phongo_binary_init_ce(INIT_FUNC_ARGS)
+			return;
+		}
+
+		float v = (float) Z_DVAL_P(val);
+
+		if (!bson_vector_float32_view_write(view, &v, 1, i)) {
+			phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "bson_vector_float32_view_write failed to write vector[%zu]", i);
+			return;
+		}
+
+		i += 1;
+	}
+	ZEND_HASH_FOREACH_END();
+
+	phongo_binary_init_vector_from_bson_key(intern, &doc, "vector");
+}
+
+static void phongo_binary_init_vector_from_int8_array(php_phongo_binary_t* intern, HashTable* vector)
+{
+	if (!zend_array_is_list(vector)) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT, "Expected vector to be a list");
+		return;
+	}
+
+	const size_t vector_len = zend_array_count(vector);
+
+	bson_t                  doc = BSON_INITIALIZER;
+	bson_vector_int8_view_t view;
+
+	if (!BSON_APPEND_VECTOR_INT8_UNINIT(&doc, "vector", vector_len, &view)) {
+		phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "BSON_APPEND_VECTOR_INT8_UNINIT failed for vector of size %zu", vector_len);
+		return;
+	}
+
+	zval*  val;
+	size_t i = 0;
+
+	ZEND_HASH_FOREACH_VAL_IND(vector, val)
+	{
+		if (Z_TYPE_P(val) != IS_LONG) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT, "Expected vector[%zu] to be an integer, %s given", i, zend_zval_type_name(val));
+			return;
+		}
+
+		if (Z_LVAL_P(val) < INT8_MIN || Z_LVAL_P(val) > INT8_MAX) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT, "Expected vector[%zu] to be a signed 8-bit integer, %" PHONGO_LONG_FORMAT " given", i, Z_LVAL_P(val));
+			return;
+		}
+
+		int8_t v = (int8_t) Z_LVAL_P(val);
+
+		if (!bson_vector_int8_view_write(view, &v, 1, i)) {
+			phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "bson_vector_int8_view_write failed to write vector[%zu]", i);
+			return;
+		}
+
+		i += 1;
+	}
+	ZEND_HASH_FOREACH_END();
+
+	phongo_binary_init_vector_from_bson_key(intern, &doc, "vector");
+}
+
+static void phongo_binary_init_vector_from_packed_bit_array(php_phongo_binary_t* intern, HashTable* vector)
+{
+	if (!zend_array_is_list(vector)) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT, "Expected vector to be a list");
+		return;
+	}
+
+	const size_t vector_len = zend_array_count(vector);
+
+	bson_t                        doc = BSON_INITIALIZER;
+	bson_vector_packed_bit_view_t view;
+
+	if (!BSON_APPEND_VECTOR_PACKED_BIT_UNINIT(&doc, "vector", vector_len, &view)) {
+		phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "BSON_APPEND_VECTOR_PACKED_BIT_UNINIT failed for vector of size %zu", vector_len);
+		return;
+	}
+
+	zval*  val;
+	size_t i = 0;
+
+	ZEND_HASH_FOREACH_VAL_IND(vector, val)
+	{
+		if (Z_TYPE_P(val) != IS_LONG && Z_TYPE_P(val) != IS_TRUE && Z_TYPE_P(val) != IS_FALSE) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT, "Expected vector[%zu] to be an integer or boolean, %s given", i, zend_zval_type_name(val));
+			return;
+		}
+
+		if (Z_TYPE_P(val) == IS_LONG && (Z_LVAL_P(val) < 0 || Z_LVAL_P(val) > 1)) {
```

Comment:
Why do you use `<` and `>` instead of `!=`?
```suggestion
		if (Z_TYPE_P(val) == IS_LONG && (Z_LVAL_P(val) != 0 || Z_LVAL_P(val) != 1)) {
```

Reply:
I've typically used `<` and `>` for range validation, as the `||` can short-circuit. This is a special case where we need only check for two values, so I can switch to:

```c
Z_TYPE_P(val) == IS_LONG && Z_LVAL_P(val) != 0 && Z_LVAL_P(val) != 1
```